1.Write a function that inputs a number and prints the multiplication table of that number

In [1]:
def number_table(n):
  for i in range(1,11):
    res = int(n) * int(i)
    print("{} * {}  : {}".format(n,i,res))
 
num = input("Enter your number: ")
print("multiplication table of the {} is :".format(num))
number_table(num)


Enter your number: 25
multiplication table of the 25 is :
25 * 1  : 25
25 * 2  : 50
25 * 3  : 75
25 * 4  : 100
25 * 5  : 125
25 * 6  : 150
25 * 7  : 175
25 * 8  : 200
25 * 9  : 225
25 * 10  : 250


2. Write a program to print twin primes less than 1000. If two consecutive odd numbers are both prime then they are known as twin primes

In [2]:
def twin_prime(lst):
  odd_list = list(map(lambda a:a,filter(lambda b: b%2 , lst)))
  print(odd_list)
  
  p_lst=[]
  val = odd_list[-1]
      
  for i in odd_list:
    #print(i)
    if i >1:
      for j in range(2,i):                      
        if (i%j)== 0:
          #print("number is not prime",i)
          break
      else:
        #print("number is prime",i)
        p_lst.append(i)
  #print(p_lst)

# to check twin odd primes number
  for j in range(0,len(p_lst)-1):
    if abs(p_lst[j] - p_lst[j + 1]) == 2.0:
      print(p_lst[j],p_lst[j+1])

  
prime_list= range(1,1000)
twin_prime(prime_list)

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99, 101, 103, 105, 107, 109, 111, 113, 115, 117, 119, 121, 123, 125, 127, 129, 131, 133, 135, 137, 139, 141, 143, 145, 147, 149, 151, 153, 155, 157, 159, 161, 163, 165, 167, 169, 171, 173, 175, 177, 179, 181, 183, 185, 187, 189, 191, 193, 195, 197, 199, 201, 203, 205, 207, 209, 211, 213, 215, 217, 219, 221, 223, 225, 227, 229, 231, 233, 235, 237, 239, 241, 243, 245, 247, 249, 251, 253, 255, 257, 259, 261, 263, 265, 267, 269, 271, 273, 275, 277, 279, 281, 283, 285, 287, 289, 291, 293, 295, 297, 299, 301, 303, 305, 307, 309, 311, 313, 315, 317, 319, 321, 323, 325, 327, 329, 331, 333, 335, 337, 339, 341, 343, 345, 347, 349, 351, 353, 355, 357, 359, 361, 363, 365, 367, 369, 371, 373, 375, 377, 379, 381, 383, 385, 387, 389, 391, 393, 395, 397, 399, 401, 403, 405, 407, 409, 411, 413, 415, 417, 419, 421,

3.Write a program to find out the prime factors of a number. Example: prime factors of 56 - 2, 2, 2, 7

In [3]:
import math
import pdb; 

def prime_factor(n):

  n = int(n)
# to check the if number is even
  while(int(n)%2 == 0):
    print(2)
    n = int(n)/2
# to check if number is odd
# pdb.set_trace()
#  print(type(n))
  for i in range(3, int(math.sqrt(n))+1, 2):
#    pdb.set_trace()
#    print(type(i))
    while(int(n)% int(i) == 0):
      print(i)
      n= int(n)/ i
# if number is itself prime number
  if n > 2:
    print(n)

num = input("Enter your number: ")
print("prime factor of the {} is :".format(num))
prime_factor(num)

Enter your number: 5
prime factor of the 5 is :
5


4. Write a program to implement these formulae of permutations and combinations. Number of permutations of n objects taken r at a time: p(n, r) = n! / (n-r)!. Number of combinations of n objects taken r at a time is: c(n, r) = n! / (r!*(n-r)!) = p(n,r) / r!

In [4]:
def fact(n):
  if (n==0 or n==1):
    return 1
  else:
    return int(n) * fact(int(n)-1)
#print(" factorial of n is", f)

def permut(num,freq):
   p = fact(int(num))/fact(int(num)-int(freq))
   print("permutation of number is :",p)

def combin(num,freq):
  c = fact(int(num))/(fact(int(freq))* fact(int(num)-int(freq)))
  print("combination of number is :",c)



n = input("Enter number for objects: ")
r = input("Enter value for r: ")
permut(n,r)
combin(n,r)






Enter number for objects: 55
Enter value for r: 6
permutation of number is : 20872566000.0
combination of number is : 28989675.0


5. Write a function that converts a decimal number to binary number

In [6]:
def dec_to_biary(n):
    a= True
    s=''
    while(a):
      q = int(n)/2
      r = int(n) % 2
      n = int(q)
      s = s + str(r)
#      print(s)
      if n == 0:
        break        
#    print(s)
    strlength=len(s) 
    res=s[strlength::-1]
    print (res)
 
num = input("Enter your number: ")
print("binary number of the {} is :".format(num))
dec_to_biary(num)

Enter your number: 5
binary number of the 5 is :
101


6. Write a function cubesum() that accepts an integer and returns the sum of the cubes of
individual digits of that number. Use this function to make functions PrintArmstrong() and
isArmstrong() to print Armstrong numbers and to find whether is an Armstrong number.

In [7]:
def cubesum(n):
    res =0
    while(int(n) > 0 ):
      r= int(n)% 10
      res = res + int(r**3)
      n = int(n) / 10
      if n ==0:
        break
    return res
  
def PrintArmstrong(x):
  x = int(x)
  y = cubesum(x)
  if (y == x):
    print('number {} is Armstrong'.format(x))
  else:
    print('number {} is not Armstrong'.format(x))

num = input("Enter your number: ")
PrintArmstrong(num)

    

Enter your number: 89
number 89 is not Armstrong


7. Write a function prodDigits() that inputs a number and returns the product of digits of that
number


In [8]:
def prodDigits(n):
    res =1
    while(int(n) > 0 ):
      r= int(n)% 10
      res = res * int(r)
      n = int(n) / 10
      if n ==0:
         break
    print(res)

num = input("Enter your number: ")
print("product of digits of number {} is :".format(num))
prodDigits(num)        


Enter your number: 89
product of digits of number 89 is :
72


8. If all digits of a number n are multiplied by each other repeating with the product, the one
digit number obtained at last is called the multiplicative digital root of n. The number of
times digits need to be multiplied to reach one digit is called the multiplicative
persistance of n.
Example: 86 -> 48 -> 32 -> 6 (MDR 6, MPersistence 3)
 341 -> 12->2 (MDR 2, MPersistence 2)
Using the function prodDigits() of previous exercise write functions MDR() and
MPersistence() that input a number and return its multiplicative digital root and
multiplicative persistence respectively

In [9]:
def prodDigits(n):
    res =1
    while(int(n) > 0 ):
      r= int(n)% 10
      res = res * int(r)
      n = int(n) / 10
      if n ==0:
         break
    return res

def count(n):
  count=0
  n = int
  while(n>0):
    count =count +1
    n= int(n)/10
#    print(n)
  return count
    


def MDR(num):
  #p= prodDigits(num)
  l= len(str(num))
  while(l>1):
    num = prodDigits(num)
    l= len(str(num))
    
  print("multiplicative digital root is",num)

def MPersistence(num):
  cont=0
  l= len(str(num))  
  while(l > 1):
    cont = cont +1
    num = prodDigits(num)
    #print(str(l) +" " + str(cont) + " " + str(num))
    l= len(str(num))
    #print(str(l) +" " + str(cont) + " " + str(num))

    #print(cont)
    #break
  print("persistence of number is",cont)

numbr = input("Enter your number: ")
MDR(numbr)
MPersistence(numbr)




Enter your number: 4
multiplicative digital root is 4
persistence of number is 0


9. Write a function sumPdivisors() that finds the sum of proper divisors of a number. Proper
divisors of a number are those numbers by which the number is divisible, except the
number itself. For example proper divisors of 36 are 1, 2, 3, 4, 6, 9, 18

In [12]:
def sumPdivisors(n):
  sum = 0
  for i in range(1,int(n)):
    if int(n)% int(i) == 0:
      sum = sum + int(i)
      print(i)
  print(sum)


num= input("Enter a number :")
sumPdivisors(num)

Enter a number :7
1
1


10. A number is called perfect if the sum of proper divisors of that number is equal to the
number. For example 28 is perfect number, since 1+2+4+7+14=28. Write a program to
print all the perfect numbers in a given range

In [11]:
def sumPdivisors(n):
  sum = 0
  for i in range(1,int(n)):
    if int(n)% int(i) == 0:
      sum = sum + int(i)
#      print(i)
  return sum

def perfect(lst):
  list2 =[]
  for i in lst:
    p = sumPdivisors(i)
    if p == i:
      list2.append(i)
  print(list2)   
      

num_list= range(1,30)
perfect(num_list)






[6, 28]


11. Two different numbers are called amicable numbers if the sum of the proper divisors of
each is equal to the other number. For example 220 and 284 are amicable numbers.
Sum of proper divisors of 220 = 1+2+4+5+10+11+20+22+44+55+110 = 284
Sum of proper divisors of 284 = 1+2+4+71+142 = 220
Write a function to print pairs of amicable numbers in a range

In [13]:
def sumPdivisors(n):
  sum = 0
  for i in range(1,int(n)):
    if int(n)% int(i) == 0:
      sum = sum + int(i)
#      print(i)
  return sum

def perfect(lst):
  lst2 =[]
  y = {v: sumPdivisors(v) for v in lst}
  print(y)
  for key, value in y.items():
    for x in lst:
      if x == value:
        lst2.append(value)
  #print(lst2)

  for key, value in y.items():
    for z in lst2:
      if y[key]==z and y[z]==key:
        print(key,z)

          
  
num_list= range(220,285)
perfect(num_list)


{220: 284, 221: 31, 222: 234, 223: 1, 224: 280, 225: 178, 226: 116, 227: 1, 228: 332, 229: 1, 230: 202, 231: 153, 232: 218, 233: 1, 234: 312, 235: 53, 236: 184, 237: 83, 238: 194, 239: 1, 240: 504, 241: 1, 242: 157, 243: 121, 244: 190, 245: 97, 246: 258, 247: 33, 248: 232, 249: 87, 250: 218, 251: 1, 252: 476, 253: 35, 254: 130, 255: 177, 256: 255, 257: 1, 258: 270, 259: 45, 260: 328, 261: 129, 262: 134, 263: 1, 264: 456, 265: 59, 266: 214, 267: 93, 268: 208, 269: 1, 270: 450, 271: 1, 272: 286, 273: 175, 274: 140, 275: 97, 276: 396, 277: 1, 278: 142, 279: 137, 280: 440, 281: 1, 282: 294, 283: 1, 284: 220}
220 284
284 220


12. Write a program which can filter odd numbers in a list by using filter function


In [14]:
def Filt_odd(num_list):
  #y = {lambda x: for x in num_list if x%2 !=0 , num_list}
  y = map(lambda v : v , filter(lambda u : u % 2, num_list))
  print("odd number list is :", list(y))
  

lst= range(1,50)
Filt_odd(lst)




odd number list is : [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49]


13. Write a program which can map() to make a list whose elements are cube of elements in
a given list

In [ ]:
def cube_num(num_list):

#  y = {lambda x: for x in num_list if x%2 !=0 , num_list}
#  y = {v: v ** 3 for v in num_list}
   y = list(map(lambda v: v ** 3, num_list ))
   print(y)
   exist_list = [i for i in y if i in num_list]
   print(exist_list)
  
lst= range(1,10)
cube_num(lst)

[1, 8, 27, 64, 125, 216, 343, 512, 729]
[1, 8]


14. Write a program which can map() and filter() to make a list whose elements are cube of
even number in a given list

In [ ]:
def cube_map_filter(num_list):
#  y = {lambda x: for x in num_list if x%2 !=0 , num_list}
  y = list(map(lambda v : v ** 3 ,num_list))
  print(y)
  x = list(map(lambda v : v , filter(lambda u : u%2==0, y)))
  print(x)

#  exist_list = [i for i in x if i in num_list]
#  print(exist_list)

  
  
lst= range(1,9)
cube_map_filter(lst)

[1, 8, 27, 64, 125, 216, 343, 512]
[8, 64, 216, 512]
